### PR TITLE
Fix compiler crash on invalid typespec

### LINKED ISSUE #12246: Invalid typespec crashes compiler
### Elixir and Erlang/OTP versions

Erlang/OTP 25 [erts-13.1.1] [source] [64-bit] [smp:8:8] [ds:8:8:10] [async-threads:1] [jit:ns]

Elixir 1.14.2 (compiled with Erlang/OTP 25)

### Operating system

any

### Current behavior

This code
```
defmodule Repro do
  @spec my_fun(integer()()) :: any()
  def my_fun(a), do: a
end
```
results in
```
== Compilation error in file lib/repro.ex ==
** (Protocol.UndefinedError) protocol String.Chars not implemented for {:integer, [line: 2], []} of type Tuple
    (elixir 1.14.2) lib/string/chars.ex:3: String.Chars.impl_for!/1
    (elixir 1.14.2) lib/string/chars.ex:22: String.Chars.to_string/1
    (elixir 1.14.2) lib/kernel/typespec.ex:857: Kernel.Typespec.typespec/4
    (stdlib 4.1.1) lists.erl:1462: :lists.mapfoldl_1/3
    (elixir 1.14.2) lib/kernel/typespec.ex:988: Kernel.Typespec.fn_args/5
    (elixir 1.14.2) lib/kernel/typespec.ex:974: Kernel.Typespec.fn_args/6
    (elixir 1.14.2) lib/kernel/typespec.ex:390: Kernel.Typespec.translate_spec/8
    (stdlib 4.1.1) lists.erl:1462: :lists.mapfoldl_1/3
```

### Expected behavior

A better error message

### PR DESCRIPTION
**Pull request by sabiwara**
Close #12246

### CODE REVIEW

File: lib/elixir/lib/kernel/typespec.ex [hunk 1]
```diff
@@ -841,7 +841,7 @@ defmodule Kernel.Typespec do
     {{:type, line(meta), :fun, args}, state}
   end
 
-  defp typespec({name, meta, args}, vars, caller, state) do
+  defp typespec({name, meta, args}, vars, caller, state) when is_atom(name) do
```
Comment (sabiwara) [inline lib/elixir/lib/kernel/typespec.ex:844]: Remote calls are already handled above, so I think it should never be anything else than an atom here?

**Comment by josevalim** [comment]
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:

**Comment by josevalim** [comment]
Looks great! I also found this one:

```
iex(1)> defmodule Foo do
...(1)> @type t :: %URI.t(){}
...(1)> end
```

Can you please take a look?

**Comment by sabiwara** [comment]
> Looks great! I also found this one:
> 
> ```
> iex(1)> defmodule Foo do
> ...(1)> @type t :: %URI.t(){}
> ...(1)> end
> ```
> 
> Can you please take a look?

Sure, will have a look!

**Pull request merged**